### DISCUSSION TITLE
SHARED_MEMORY_SERVER support

### DISCUSSION THREAD

**Discussion by YunfeiBai** [bulletphysics/bullet3#3756] [General]
Having an equivalent version of the function below for headless mode, to start an in-process physics server and physics client in process with shared memory, which PyBullet can connect to
https://github.com/bulletphysics/bullet3/blob/a506879d6cfd44f7df041d4fd05b3641fd1ed713/examples/SharedMemory/SharedMemoryInProcessPhysicsC_API.h#L14

**Comment by mbusy** [comment]
Hi, any updates on the topic ?

**Comment by erwincoumans** [comment]
pip3 install pybullet -u
python3 -m pybullet_utils.runServer&
Is that what you were looking for?

**Comment by mbusy** [comment]
Hi Erwin, more or less yes! (without a GUI, but I'm guessing that I can use something like pybullet.connect(SHARED_MEMORY_SERVER) to do that).

I was actually wondering if creating several physics servers that way would allow me to run multiple simulations in parallel (using multiple cores)? Thanks again !

**Comment by erwincoumans** [comment]
To run many simulations in parallel on many cpu cores, I recommend using pubullet_utils.bullet_client and multiprocessing like this example.

https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/gym/pybullet_robots/panda/batchsim3_grasp.py

**Comment by mbusy** [comment]
Thanks!

**Comment by erwincoumans** [comment]
Sure. One more example using MPI instead with pybullet_client is deep_mimic:
python3 mpi_run.py --arg_file train_humanoid3d_spinkick_args.txt --num_workers 4
mpiexec -n 16 python3 -m pybullet_envs.deep_mimic.DeepMimic_Optimizer

This latter is rather finicky, since Tensorflow and MPI sometimes break their API.
